Save linked subresources when saving web page resources

szewai opened the pull request:
#### f006d3e24eebe0b6ca75dc65c939d9ce182af932
<pre>
Save linked subresources when saving web page resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=267372">https://bugs.webkit.org/show_bug.cgi?id=267372</a>
<a href="https://rdar.apple.com/120491493">rdar://120491493</a>

Reviewed by Ryosuke Niwa and BJ Burg.

When collecting subresources URLs, we should include URL of link element if it is a link to external resource
(<a href="https://html.spec.whatwg.org/multipage/links.html#linkTypes)">https://html.spec.whatwg.org/multipage/links.html#linkTypes)</a>, and it may fetch or cache the linked resource
(<a href="https://html.spec.whatwg.org/multipage/semantics.html#fetching-and-processing-a-resource-from-a-link-element).">https://html.spec.whatwg.org/multipage/semantics.html#fetching-and-processing-a-resource-from-a-link-element).</a> We
currently only include URL of style sheet.

Update API test: WebArchive.SaveResourcesLink

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::mayFetchResource):
(WebCore::HTMLLinkElement::addSubresourceAttributeURLs const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/272925@main">https://commits.webkit.org/272925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b82ed31e01bfc383719fe2c096ad7c23e46f2ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33634 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12408 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36252 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34680 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9565 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34110 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10461 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29988 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9138 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37577 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30317 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9366 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/7321 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/33250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11144 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7772 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9950 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -615,17 +615,28 @@ std::optional<LinkIconType> HTMLLinkElement::iconType() const
 }
 
+// https://html.spec.whatwg.org/multipage/links.html#linkTypes
+static bool isLinkedToExternalResource(LinkRelAttribute relAttribute)
```
Naming suggestion: `linksToExternalResource(attr)`

szewai replied:
Based on @rniwa's comment, It seems we only need to collect URLs for subresources that may be fetched based on https://html.spec.whatwg.org/multipage/semantics.html#fetching-and-processing-a-resource-from-a-link-element, so I updated this to `mayFetchResource`. Let me know if you think there is better naming for this.

rniwa commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -615,17 +615,28 @@ std::optional<LinkIconType> HTMLLinkElement::iconType() const
+static bool isLinkedToExternalResource(LinkRelAttribute relAttribute)
+{
+    return relAttribute.isDNSPrefetch
+    || relAttribute.isStyleSheet
```
Nit: || should be indented by 4 spaces.

rniwa commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -615,17 +615,28 @@ std::optional<LinkIconType> HTMLLinkElement::iconType() const
+// https://html.spec.whatwg.org/multipage/links.html#linkTypes
+static bool isLinkedToExternalResource(LinkRelAttribute relAttribute)
+{
+    return relAttribute.isDNSPrefetch
```
Surely, we don't need to include DNSPrefetch?

rniwa commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -615,17 +615,28 @@ std::optional<LinkIconType> HTMLLinkElement::iconType() const
+    || relAttribute.isLinkModulePreload
+    || relAttribute.isLinkPreload
+    || relAttribute.isLinkPrefetch
+    || relAttribute.isLinkPreconnect
```
We shouldn't need to do prefetch or preconnect either

szewai replied:
It seems prefetch could request the resource (`LinkLoader::prefetchIfNeeded`) but preconnect and DNSPrefetch do not, so I removed reconnect and DNSPrefetch from the list.

rniwa replied:
Well, prefetch COULD request the resource but the effect isn't script observable so we don't need to do it.

szewai replied:
Ok, removed

rniwa commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -615,17 +615,26 @@ std::optional<LinkIconType> HTMLLinkElement::iconType() const
+    return relAttribute.isStyleSheet
+        || relAttribute.isLinkModulePreload
+        || relAttribute.isLinkPreload
+        || relAttribute.isLinkPrefetch
```
I don't think we need to include prefetch either.

rniwa reviewed (approved):
r=me with comments.

szewai committed:
Save linked subresources when saving web page resources
https://bugs.webkit.org/show_bug.cgi?id=267372
rdar://120491493

Reviewed by Ryosuke Niwa and BJ Burg.

When collecting subresources URLs, we should include URL of link element if it is a link to external resource
(https://html.spec.whatwg.org/multipage/links.html#linkTypes), and it may fetch or cache the linked resource
(https://html.spec.whatwg.org/multipage/semantics.html#fetching-and-processing-a-resource-from-a-link-element). We
currently only include URL of style sheet.

Update API test: WebArchive.SaveResourcesLink

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::mayFetchResource):
(WebCore::HTMLLinkElement::addSubresourceAttributeURLs const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/272925@main
